flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryService.java

```diff
@@ -0,0 +1,363 @@
+			.filter(p -> factoryClass.isAssignableFrom(p.getClass()))
+			.collect(Collectors.toList());
+
+		if (classFactories.isEmpty()) {
+			throw new NoMatchingTableFactoryException(
+				String.format("No factory implements '%s'.", factoryClass.getCanonicalName()),
+				factoryClass,
+				foundFactories,
+				properties);
+		}
+
+		return classFactories;
+	}
+
+	/**
+	 * Filters for factories with matching context.
+	 *
+	 * @return all matching factories
+	 */
+	private static <T> List<TableFactory> filterByContext(
+		Class<T> factoryClass,
+		Map<String, String> properties,
+		List<TableFactory> foundFactories,
+		List<TableFactory> classFactories) {
+
+		List<TableFactory> matchingFactories = classFactories.stream().filter(factory -> {
+			Map<String, String> requestedContext = normalizeContext(factory);
+
+			Map<String, String> plainContext = new HashMap<>(requestedContext);
+			// we remove the version for now until we have the first backwards compatibility case
+			// with the version we can provide mappings in case the format changes
+			plainContext.remove(CONNECTOR_PROPERTY_VERSION);
+			plainContext.remove(FORMAT_PROPERTY_VERSION);
+			plainContext.remove(METADATA_PROPERTY_VERSION);
+			plainContext.remove(STATISTICS_PROPERTY_VERSION);
+			plainContext.remove(CATALOG_PROPERTY_VERSION);
+
+			// check if required context is met
+			return plainContext.keySet().stream().allMatch(e -> properties.containsKey(e) && properties.get(e).equals(plainContext.get(e)));
+		}).collect(Collectors.toList());
+
+		if (matchingFactories.isEmpty()) {
+			throw new NoMatchingTableFactoryException(
+				"No context matches.",
+				factoryClass,
+				foundFactories,
+				properties);
+		}
+
+		return matchingFactories;
+	}
+
+	/**
+	 * Prepares the properties of a context to be used for match operations.
+	 */
+	private static Map<String, String> normalizeContext(TableFactory factory) {
+		Map<String, String> requiredContext = factory.requiredContext();
+		if (requiredContext == null) {
+			throw new TableException(
+				String.format("Required context of factory '%s' must not be null.", factory.getClass().getName()));
+		}
+		return requiredContext.keySet().stream()
+			.collect(Collectors.toMap(key -> key.toLowerCase(), key -> requiredContext.get(key)));
+	}
+
+	/**
+	 * Filters the matching class factories by supported properties.
+	 */
+	private static <T> T filterBySupportedProperties(
+		Class<T> factoryClass,
+		Map<String, String> properties,
+		List<TableFactory> foundFactories,
+		List<TableFactory> classFactories) {
+
+		final List<String> plainGivenKeys = new LinkedList<>();
+		properties.keySet().forEach(k -> {
+			// replace arrays with wildcard
+			String key = k.replaceAll(".\\d+", ".#");
+			// ignore duplicates
+			if (!plainGivenKeys.contains(key)) {
+				plainGivenKeys.add(key);
+			}
+		});
+
+		AtomicReference<Optional<String>> lastKey = new AtomicReference<>(Optional.empty());
```

Comment:
Let's rewrite this as for-loop if we need additional return from the loop. I had hard time understanding why do we need `AtomicReference` here.